### ISSUE TITLE
Scheduler occasionally has stale state and does not assign pods to nodes that could fit them

### ISSUE THREAD

**Issue by javidiaz** [kubernetes/kubernetes#102339]
We have a Kubernetes cluster with version v1.19.7.  We have observed that sometimes the kubernetes scheduler does not schedule pods on nodes that are available for them, and the pods remain pending. We added some custom logs to the scheduler code and we were able to determine that when this happens, the scheduler thinks that there is a pod on the available node, even though that pod no longer exists in the cluster.

#### What happened:

From our debugging, here’s the sequence of events we have observed:

1. A pause pod P is running on an existing node N in a certain node pool. This pause pod exists simply to overprovision the node pool.
2. A new pod X, which takes up most of a node, preempts the pause pod P.
3. The new pod X is deleted as part of a scale down, before it is bound to node N.
4. The node N is removed from the cluster (either scaled down because of underutilization or preempted).
5. The scheduler removes the node N from the node tree.
6. Later, a new node is added to the cluster, with the same name N as the previous node. It could be in the same or different node pool.
7. The scheduler adds the new node N to the node tree.
8. The autoscaler (correctly) decides that a pending pod Y can be scheduled on the node N as it has the resources and taints to fit the pod.
9. The scheduler does not schedule the pod Y on the node N. From our debug logs, we discovered that in such cases, the scheduler thinks pod X is still on the node, even though the pod was never bound to the node, and the node has since been removed and re-added to the cluster ([this is where we added this log statement](https://github.com/kubernetes/kubernetes/blob/v1.19.7/pkg/scheduler/framework/plugins/noderesources/fit.go#L226)).

`CUSTOMLOG: Pod scaletest-5b9589-mbm4f requests 13000 cpu, 62277025792 memory (13000 cpu, 62277025792 memory). Node c-10-92-64-72.cloud.domain.com has 6 pods, 26161/15000 used/allocatable cpu, 135285178368/66375368704 used/allocatable memory. Pods in node: admin-namespace/node-problem-detector-cpk5f/00621b52-fc6e-4786-90d8-c9db3ed0877c,gcp-csi-namespace/csi-gce-pd-node-xt8c8/3d172e2b-572d-4497-8098-6e63c7002823,kube-system/kp-c-10-92-64-72.cloud.domain.com/b4af72c6-6306-4006-b8ee-03077312c8e2,admin-namespace/fluent-bit-node-omega-vswwn/82c144df-6b29-4d03-a79d-c1541d71f7a6,datadog-namespace/datadog-agent-zzl2m/7a79591f-5115-44ef-b980-b0fa6a051d40,mynamespace/podX/bb59f18d-7ad5-455e-a4e8-61525ac81f52. InsufficientResources?: Insufficient cpu,Insufficient memory`

From the node’s describe:

```
Non-terminated Pods:          (5 in total)
 Namespace         Name                              CPU Requests CPU Limits  Memory Requests Memory Limits AGE
 ---------         ----                              ------------ ----------  --------------- ------------- ---
 admin-namespace   fluent-bit-node-omega-vswwn       501m (3%)    1501m (10%) 818Mi (1%)      1074Mi (1%)   28m
 admin-namespace   node-problem-detector-cpk5f       0 (0%)       0 (0%)      0 (0%)          0 (0%)        28m
 kube-system       kp-c-10-92-64-72.cloud.domain.com 500m (3%)    500m (3%)   2Gi (3%)        2Gi (3%)      27m
 datadog-namespace datadog-agent-zzl2m               60m (0%)     10m (0%)    1Gi (1%)        64Mi (0%)     28m
 gcp-csi-namespace csi-gce-pd-node-xt8c8             0 (0%)       0 (0%)      0 (0%)          0 (0%)        28m
```

Notice that the rest of the pods on the node (according to kubectl describe) match what the scheduler thinks is on the node; only the last pod is different. It is present on the node according to the scheduler, but not according to kubectl describe.

10. Restarting the scheduler makes it so that the pod Y is correctly scheduled on node N.

#### What you expected to happen:
Pod to be scheduled on the node without needing to restart the scheduler.

#### How to reproduce it (as minimally and precisely as possible):
The issue might be hard to reproduce since it appears to be a race condition, but here are the steps based on what we have observed:

1. Have a cluster with pod preemption enabled.
2. Have a pod with a low priority scheduled on a node.
3. Have a pod with a high priority preempt the pod with the low priority (which, if still present, would be too large to allow the later pod to be scheduled).
4. Delete the high priority pod before it is actually scheduled on the node. We believe the nominated node name is set, but not the node name, but don’t have proof of this.
5. Delete the node from the cluster (this might be optional), then add it back in with the same name.
6. Have a new pod that can be scheduled on the node.

#### Anything else we need to know?:
Cluster Configuration
We are running our own deployment of Kubernetes v1.19.7 on GCP (not using GKE). It scales from 250 nodes to 1200 nodes (average around 600 nodes). Pod count ranges from 2k to 7k (average around 4.5k pods). We have two clusters with this setup, and we observe the issue in both.

Node Pool Configuration
We have different node pools (with multiple managed instance groups for each). Our node names are based on the node’s IP. With nodes scaling up and down in the cluster, node names are often reused, so it is common for a node to disappear from one node pool and eventually reappear in a different one.

Kube-Scheduler Configuration
The configuration of kube-scheduler is:
```
    --kubeconfig=/etc/kubernetes/scheduler-kubeconfig.yaml \
    --v=4 \
    --logtostderr=false \
    --skip-log-headers=true \
    --log-file=/srv/k8s/logs/scheduler.log \
    --address=0.0.0.0 \
    --leader-elect=true \
    --leader-elect-lease-duration 65s \
    --leader-elect-renew-deadline 60s \
    --kube-api-qps=100 \
```

Autoscaler Configuration
We run a version of the open-source [autoscaler](https://github.com/kubernetes/autoscaler). We also use pause pods with negative priority to enable faster scale up for some workloads. We have not seen the issue before we started using pause pods.

/sig scheduling
/cc @alculquicondor

**Comment by ahg-g** [comment]
It seems that `podX` is not being deleted from the `nominatedPodMap`.

When the scheduler receives a delete event for a pod that is not scheduled yet, it deletes it from the scheduling queue [here](https://github.com/kubernetes/kubernetes/blob/1dd5338295409edcfff11505e7bb246f0d325d15/pkg/scheduler/eventhandlers.go#L206), which is supposed to delete it from the nominatedPodMap [here](https://github.com/kubernetes/kubernetes/blob/1dd5338295409edcfff11505e7bb246f0d325d15/pkg/scheduler/internal/queue/scheduling_queue.go#L471).

Did you notice any other errors in the scheduler logs for podX? If you are able to reproduce, it would be great if you can increase the log level to 3, and add log messages to [before this line](https://github.com/kubernetes/kubernetes/blob/1dd5338295409edcfff11505e7bb246f0d325d15/pkg/scheduler/internal/queue/scheduling_queue.go#L471) and [before this one](https://github.com/kubernetes/kubernetes/blob/1dd5338295409edcfff11505e7bb246f0d325d15/pkg/scheduler/internal/queue/scheduling_queue.go#L588) to check if the pod is actually getting deleted.

**Comment by Huang-Wei** [comment]
@javidiaz when this problem surfaced, did you spot repeated log entries with "Attempting to schedule pod: mynamespace/podX" and afterwards "Failed to bind pod: mynamespace/podX"?

@ahg-g I'm suspicious the pod deletion did happen and cleared the entry in `nominatedPodMap`, but there is somehow a phantom pod (podX) that keeps retrying and filling `nominatedPodMap`. We used to observe this symptom in a flaky integration test (PreemptionRaces), and https://github.com/kubernetes/kubernetes/pull/96071#issuecomment-720955648 was my analysis, and the fix wasn't chery-picked to 1.19. But I'm not quite sure it's related. Let's wait for @javidiaz 's response first.

**Comment by alculquicondor** [comment]
If a new workload Pod with the same name is created, would that match https://github.com/kubernetes/kubernetes/pull/96071#issuecomment-720955648?

**Comment by ahg-g** [comment]
It doesn't seem to me that podX was recreated with the same name.

**Comment by javidiaz** [comment]
Thanks for looking into this. @ahg-g we can reproduce, should I add the logs entries into those places you suggest? We have log level 4 right now.

@Huang-Wei we do not see the "Failed to bind pod.." log. These are the logs we see: 

>  May 20th 2021, 05:53:52.319 leader-scheduler Attempting to schedule pod: mynamespace/podX
 May 20th 2021, 05:53:52.319 inactive-scheduler-2 add event for unscheduled pod mynamespace/podX
 May 20th 2021, 05:53:52.319 leader-scheduler About to try and schedule pod mynamespace/podX
 May 20th 2021, 05:53:52.319 inactive-scheduler-1 add event for unscheduled pod mynamespace/podX
 May 20th 2021, 05:53:52.319 leader-scheduler add event for unscheduled pod mynamespace/podX
 May 20th 2021, 05:53:52.330 leader-scheduler "Unable to schedule pod; no fit; waiting" pod="mynamespace/podX" err="0/840 nodes are available: 1 node(s) had taint {status: maintenance}, that the pod didn't tolerate, 101 node(s) were unschedulable, 12 Too many pods, 154 Insufficient memory, 45 node(s) didn't match node selector, 693 Insufficient cpu."
 May 20th 2021, 05:53:52.331 leader-scheduler Updating pod condition for mynamespace/podX to (PodScheduled==False, Reason=Unschedulable)
 May 20th 2021, 05:53:52.354 leader-scheduler Attempting to schedule pod: mynamespace/podX
 May 20th 2021, 05:53:52.354 leader-scheduler About to try and schedule pod mynamespace/podX
 May 20th 2021, 05:53:52.359 inactive-scheduler-1 delete event for unscheduled pod mynamespace/podX
 May 20th 2021, 05:53:52.359 inactive-scheduler-2 delete event for unscheduled pod mynamespace/podX
 May 20th 2021, 05:53:52.362 leader-scheduler delete event for unscheduled pod mynamespace/podX
 May 20th 2021, 05:53:52.364 leader-scheduler Status after running PostFilter plugins for pod mynamespace/podX: &{1 [pods "podX" not found]}
 May 20th 2021, 05:53:52.364 leader-scheduler Error getting the updated preemptor pod object: pods "podX" not found
 May 20th 2021, 05:53:52.365 leader-scheduler Pod mynamespace/podX doesn't exist in informer cache: pod "podX" not found
 May 20th 2021, 05:53:52.365 leader-scheduler "Unable to schedule pod; no fit; waiting" pod="mynamespace/podX" err="0/840 nodes are available: 1 node(s) had taint {status: maintenance}, that the pod didn't tolerate, 101 node(s) were unschedulable, 12 Too many pods, 154 Insufficient memory, 45 node(s) didn't match node selector, 693 Insufficient cpu."
 May 20th 2021, 05:53:52.365 leader-scheduler Updating pod condition for mynamespace/podX to (PodScheduled==False, Reason=Unschedulable)

**Comment by Huang-Wei** [comment]
@javidiaz Thanks for the logs. They're very helpful.

I think the problem lies in the second time, when a preemptor (podX) is retried in scheduler while also being deleted elsewhere:

```
May 20th 2021, 05:53:52.354 leader-scheduler Attempting to schedule pod: mynamespace/podX
May 20th 2021, 05:53:52.362 leader-scheduler delete event for unscheduled pod mynamespace/podX
```

In this case, running PostFilter again will fail due to enable to obtain it:

```
May 20th 2021, 05:53:52.364 leader-scheduler Error getting the updated preemptor pod object: pods "podX" not found
May 20th 2021, 05:53:52.365 leader-scheduler Pod mynamespace/podX doesn't exist in informer cache: pod "podX" not found
```

After that, the error is recorded, but unfortunately, the [nominatedPod is set unconditionally](https://github.com/kubernetes/kubernetes/blob/46a4cc97ea15ebc888501f6682a6970bda53063d/pkg/scheduler/scheduler.go#L361-L367) and hence persisted the phantom pod into scheduler snapshot. That seems to be the root cause of the problem.

I will work on a fix.

/assign

**Comment by Huang-Wei** [comment]
@javidiaz if I guess it right, you should spot some log entries with "Error updating pod mynamespace/podX" repeatedly after the timestamp "May 20th 2021, 05:53:52.365"?

**Comment by javidiaz** [comment]
@Huang-Wei  Great you got a lead to fix the issue. I do not see the error log you mention for the PodX specifically, but I do see it for other replicas of the same Deployment that created PodX. Not sure if we lost the error logs for PodX or if it is stuck in some weird state that prevents that log from showing up.

The log I see is something like this 

> Error updating pod mynamespace/podX-replica: pods "podX-replica" not found

**Comment by Huang-Wei** [comment]
@javidiaz the code fix has been merged into `master` branch, and the following PRs have been created to the supported versions:

- 1.21 #102667
- 1.20 #102674
- 1.19 #102675

Once the cherry-pick PRs get merged, they will be packaged in the next Kubernetes patch respectively.

**Comment by javidiaz** [comment]
> @javidiaz the code fix has been merged into `master` branch, and the following PRs have been created to the supported versions:
> 
> * 1.21 #102667
> * 1.20 #102674
> * 1.19 #102675
> 
> Once the cherry-pick PRs get merged, they will be packaged in the next Kubernetes patch respectively.

Thanks @Huang-Wei for the quick fix!
